### PR TITLE
Prevent duplicate mentee applications

### LINKED ISSUE #227: Prevent mentees from applying for the same program twice
**Is your feature request related to a problem? Please describe.**
In the sx-february edition, we are not allowing mentees to apply for multiple mentors

**Describe the solution you'd like**
Validate the related endpoint:
`applyAsMentee` in `src/main/java/org/sefglobal/scholarx/service/MentorService.java`

**Describe alternatives you've considered**
N/A

**Additional context**
You can update this code segment to solve the issue
```
        Optional<Mentee> duplicateMentee = menteeRepository.findByProfileIdAndAppliedMentorId(profileId, mentorId);
        if (duplicateMentee.isPresent()) {
            String msg = "Error, Unable to apply as a mentee. " +
                         "Profile with id: " + profileId + " has already applied for the selected mentor.";
            log.error(msg);
            throw new BadRequestException(msg);
        }
```

### PR DESCRIPTION
**Pull request by cnonnet**
## Purpose

The purpose of this PR is to fix #227  

## Goals

Prevent mentees from applying twice for the same program

## Approach

Added the relevant validations

### Screenshots

N/A

##  Checklist
- [x] This PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.
- [x] I have read and understood the development best practices guidelines ( http://bit.ly/sef-best-practices )
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation

## Related PRs
 
N/A

## Learning

N/A

### CODE REVIEW

File: src/main/java/org/sefglobal/scholarx/service/MentorService.java [hunk 1]
```diff
@@ -138,6 +138,14 @@ public Mentee applyAsMentee(long mentorId, long profileId, Mentee mentee)
             throw new BadRequestException(msg);
         }
 
+        Optional<Mentee> alreadyApplied = menteeRepository.findByProgramIdAndProfileId(profileId, profileId);
```
Comment (Piumal1999) [inline src/main/java/org/sefglobal/scholarx/service/MentorService.java:141]:
```suggestion
        Optional<Mentee> alreadyAppliedMentee = menteeRepository.findByProgramIdAndProfileId(optionalMentor.get().getProgram().getId(), profileId);
```

File: src/main/java/org/sefglobal/scholarx/service/MentorService.java [hunk 2]
```diff
@@ -138,6 +138,14 @@ public Mentee applyAsMentee(long mentorId, long profileId, Mentee mentee)
...
+        if (alreadyApplied.isPresent()) {
```
Comment (Piumal1999) [inline src/main/java/org/sefglobal/scholarx/service/MentorService.java:142]:
```suggestion
        if (alreadyAppliedMentee.isPresent()) {
```

**Review by Piumal1999 — CHANGES_REQUESTED** [review]
Good job @cnonnet . Thank you very much

Btw, I requested some minor changes. Could you please update the Pull request with those changes.

Also Remove this part from the code. (MentorService file: Line 133 to 140)
```
        Optional<Mentee> duplicateMentee = menteeRepository.findByProfileIdAndAppliedMentorId(profileId, mentorId);
        if (duplicateMentee.isPresent()) {
            String msg = "Error, Unable to apply as a mentee. " +
                         "Profile with id: " + profileId + " has already applied for the selected mentor.";
            log.error(msg);
            throw new BadRequestException(msg);
        }
```

Also  update line 90 of `src/test/java/org/sefglobal/scholarx/service/MentorServiceTest.java` file to this:

```
                .findByProgramIdAndProfileId(anyLong(), anyLong());
```
So the build will pass

File: src/test/java/org/sefglobal/scholarx/service/MentorServiceTest.java [hunk 3]
```diff
@@ -87,7 +87,7 @@ void applyAsMentee_withValidData_thenReturnCreatedData()
                 .findByProfileIdAndProgramId(anyLong(), anyLong());
         doReturn(Optional.empty())
                 .when(menteeRepository)
-                .findByProfileIdAndAppliedMentorId(anyLong(), anyLong());
+                .findByProgramIdAndProfileId(anyLong(), anyLong())
```
Comment (Piumal1999) [inline src/test/java/org/sefglobal/scholarx/service/MentorServiceTest.java:90]:
```suggestion
                .findByProgramIdAndProfileId(anyLong(), anyLong());
```
@cnonnet  I think you missed the semi colon.

**Comment by cnonnet** [comment]
Thanks a lot for your review, obviously it was not working properly, but with your corrections, it should be better.
Sry for the failed attempt, I am fairly new to github.

**Comment by cnonnet** [comment]
Thanks for you time and your patience with me :)

**Review by Piumal1999 — APPROVED** [review]
The build is passing!! Well done @cnonnet. It's a pleasure to assist you. 🙌

If you like, you can contribute to the [sef-site](https://github.com/sef-global/sef-site) repo as well. Thank you

**Pull request merged**